Parser/Token.cs

```diff
@@ -1,53 +1,59 @@
-   public Tokenizer Source { get; }
-   public E Kind { get; }
-   public string Text { get; }
-   public int Line { get; }
-   public int Column { get; }
+public class Token {
+   public Token (Tokenizer source, E kind, string text, int line, int column, string msg)
+      => (Source, Kind, Text, Line, Column, ErrorMessage) = (source, kind, text, line, column, ErrorMessage + msg);
+   public Tokenizer Source { get; }
+   public E Kind { get; }
+   public string Text { get; }
+   public int Line { get; }
+   public int Column { get; }
+   public string ErrorMessage { get; } = "";
 
    // The various types of token
-   public enum E {
-      // Keywords
-      PROGRAM, VAR, IF, THEN, WHILE, ELSE, FOR, TO, DOWNTO,
-      DO, BEGIN, END, PRINT, TYPE, NOT, OR, AND, MOD, _ENDKEYWORDS,
-      // Operators
-      ADD, SUB, MUL, DIV, NEQ, LEQ, GEQ, EQ, LT, GT, ASSIGN, 
-      _ENDOPERATORS,
-      // Punctuation
-      SEMI, PERIOD, COMMA, OPEN, CLOSE, COLON, 
-      _ENDPUNCTUATION,
-      // Others
-      IDENT, INTEGER, REAL, BOOLEAN, STRING, CHAR, EOF, ERROR
+   public enum E {
+      // Keywords
+      PROGRAM, VAR, IF, THEN, WHILE, ELSE, FOR, TO, DOWNTO,
+      DO, BEGIN, END, PRINT, TYPE, NOT, OR, AND, MOD, _ENDKEYWORDS,
+      // Operators
+      ADD, SUB, MUL, DIV, NEQ, LEQ, GEQ, EQ, LT, GT, ASSIGN,
+      _ENDOPERATORS,
+      // Punctuation
+      SEMI, PERIOD, COMMA, OPEN, CLOSE, COLON,
+      _ENDPUNCTUATION,
+      // Others
+      IDENT, INTEGER, REAL, BOOLEAN, STRING, CHAR, EOF, ERROR
    }
 
    // Print a Token
-   public override string ToString () => Kind switch {
-      EOF or ERROR => Kind.ToString (),
-      < _ENDKEYWORDS => $"\u00ab{Kind.ToString ().ToLower ()}\u00bb",
-      STRING => $"\"{Text}\"",
-      CHAR => $"'{Text}'",
-      _ => Text,
+   public override string ToString () => Kind switch {
+      EOF or ERROR => Text,
+      < _ENDKEYWORDS => $"\u00ab{Kind.ToString ().ToLower ()}\u00bb",
+      STRING => $"\"{Text}\"",
+      CHAR => $"'{Text}'",
+      _ => Text,
    };
 
    // Utility function used to echo an error to the console
-   public void PrintError () {
-      if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
-      Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
-      Console.ResetColor ();
-   }
+   public void PrintError (bool isLast) {
+      if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
+      if (mY != Line) (mX, mY) = (0, Line);
+      mX = Math.Max (mX, Column - ErrorMessage.Length / 2) + 1 + mX;
+      Console.ForegroundColor = ConsoleColor.Yellow;
+      Console.WriteLine ("^", Console.CursorLeft = Column);
+      Console.WriteLine ($"{ErrorMessage}", Console.CursorLeft = mX);
+      Console.ResetColor ();
+      Console.CursorTop = isLast ? Console.CursorTop : Console.CursorTop - 2;
```

Comment:
The error message should have the above 2 lines (if any), the error line with the message and 2 lines below it (if any).